tests/Mremi/UrlShortener/Tests/Provider/Wechat/WechatProviderTest.php

```diff
@@ -0,0 +1,305 @@
+     */
+    public function testShortenThrowsExceptionIfApiResponseIsString()
+    {
+        $this->mockClient($this->getMockResponseAsString());
+
+        $this->provider->shorten($this->getBaseMockLink());
+    }
+
+    /**
+     * Tests the shorten method throws exception if Wechat returns a response with no errcode.
+     *
+     * @expectedException        \Mremi\UrlShortener\Exception\InvalidApiResponseException
+     * @expectedExceptionMessage Property "errcode" does not exist within Wechat response.
+     */
+    public function testShortenThrowsExceptionIfApiResponseHasNoErrorCode()
+    {
+        $this->mockClient($this->getMockResponseAsInvalidObject());
+
+        $this->provider->shorten($this->getBaseMockLink());
+    }
+
+    /**
+     * Tests the shorten method throws exception if Wechat returns an invalid status code.
+     *
+     * @expectedException        \Mremi\UrlShortener\Exception\InvalidApiResponseException
+     * @expectedExceptionMessage Wechat returned status code "40001" with message "invalid credential, access_token is invalid or not latest hint: [miSs30226vr31!]"
+     */
+    public function testShortenThrowsExceptionIfApiResponseHasInvalidStatusCode()
+    {
+        $this->mockClient($this->getMockResponseWithInvalidStatusCode());
+
+        $this->provider->shorten($this->getBaseMockLink());
+    }
+
+    /**
+     * Tests the shorten method with a valid Wechat's response.
+     */
+    public function testShortenWithValidApiResponse()
+    {
+        $response = $this->getBaseMockResponse();
+
+        $apiRawResponse = <<<'JSON'
+{
+    "errcode": 0,
+    "errmsg": "ok",
+    "short_url": "https://w.url.cn/s/ATYfzFm"
+}
+JSON;
+
+        $stream = $this->getBaseMockStream();
+        $stream
+            ->expects($this->once())
+            ->method('getContents')
+            ->will($this->returnValue($apiRawResponse));
+
+        $response
+            ->expects($this->once())
+            ->method('getBody')
+            ->will($this->returnValue($stream));
+
+        $link = $this->getMockLongLink();
+        $link
+            ->expects($this->once())
+            ->method('setShortUrl')
+            ->with($this->equalTo('https://w.url.cn/s/ATYfzFm'));
+
+        $this->mockClient($response);
+
+        $this->provider->shorten($link);
+    }
+
+    /**
+     * Tests the expand method throws exception if Wechat returns a string.
+     *
+     * @expectedException        \Mremi\UrlShortener\Exception\InvalidApiResponseException
+     * @expectedExceptionMessage Wechat does not support expand url yet.
+     */
+    public function testExpandThrowsExceptionIfApiResponseIsString()
+    {
+        //$this->mockClient($this->getMockResponseAsString());
+
+        $this->provider->expand($this->getBaseMockLink());
+    }
+
+    /**
+     * Initializes the provider.
+     */
+    protected function setUp()
```

Comment:
Can you move it at the top of the class please (same for `tearDown`)? By convention, these methods are declared first to increase readability.